apps/src/blockly/addons/cdoSpritePointer.js

```diff
@@ -0,0 +1,112 @@
+ * @returns The url of the image that the pointer block should display, or an empty string if the block should display no image
+ */
+export function getPointerBlockImageUrl(
+  block,
+  pointerMetadataMap,
+  imageSourceId
+) {
+  const pointerData = pointerMetadataMap[block.type];
+  if (!pointerData || !block.inputList || block.inputList.length === 0) {
+    return '';
+  }
+  let imageSource = undefined;
+  const mainWorkspace = Blockly.getMainWorkspace();
+  if (imageSourceId !== undefined) {
+    imageSource = mainWorkspace.getBlockById(imageSourceId);
+  }
+  if (!imageSource) {
+    const rootBlock = block.getRootBlock();
+    if (rootBlock.type === pointerData.imageSourceType) {
+      imageSource = rootBlock;
+    } else if (block.getRootBlock().type === block.type) {
+      const blockWorkspace = Blockly.Workspace.getById(block.workspace.id);
+      // If this block has itself as a root and is in a flyout workspace,
+      // it is a in a mini toolbox. A block can't be moved from one flyout to
+      // another, so if it's in a flyout, it is in its original flyout.
+      if (blockWorkspace && blockWorkspace.isFlyout && block.imageSourceId) {
+        imageSource = mainWorkspace.getBlockById(block.imageSourceId);
+      }
+    }
+  }
+  if (imageSource) {
+    return getImageUrlFromImageSource(imageSource, pointerData.imageIndex);
+  } else {
+    // The block is probably disconnected from any root or toolbox. Reset to
+    // default text.
+    return '';
+  }
+}
+
+// Given a shadow block and an optional image source block id, update the image displayed
+// on the pointer block to match the image selected in the image source block. See getPointerBlockImageUrl
+// for details on how we find the image source block.
+export function updatePointerBlockImage(
+  block,
+  pointerMetadataMap,
+  imageSourceId
+) {
+  const url = getPointerBlockImageUrl(block, pointerMetadataMap, imageSourceId);
+  changePointerImage(url, block);
+}
+
+// Get the image url from the image source block's inputList at the given index.
+// We find the image url by looking at the connection to the input at the given index.
+// If the image source does not have an image at the given index, return an empty string.
+function getImageUrlFromImageSource(imageSourceBlock, imageIndexOnSource) {
+  const targetConnection =
+    imageSourceBlock.inputList[imageIndexOnSource]?.connection
+      ?.targetConnection;
+  // We only want to get the image from a connection block that is not an insertion marker.
+  // If the block is an insertion marker that means the block is being dragged
+  // and is not yet connected to the image source block.
+  if (
+    targetConnection &&
+    targetConnection.sourceBlock_ &&
+    targetConnection.sourceBlock_.type === 'gamelab_allSpritesWithAnimation' &&
+    !targetConnection.sourceBlock_.isInsertionMarker()
+  ) {
+    return (
+      targetConnection.sourceBlock_.inputList[0].fieldRow[0].imageElement_.getAttribute(
```

Comment:
Can we add a comment explaining that this works because the `'gamelab_allSpritesWithAnimation'` block only has one input with one field (the costume picker dropdown)?

Reply:
done!